src/CoilMaps.jl

```diff
@@ -1,32 +1,59 @@
+function applyDensityCompensation!(dataU::AbstractArray{Complex{T}}, trj::Matrix{T}; density_compensation::Union{Symbol, <:AbstractVector{<:AbstractVector{T}}}=:radial_3D) where {T}
+    if density_compensation == :radial_3D
+            dataU .*= transpose(sum(abs2, trj,dims=1))
+    elseif density_compensation == :radial_2D
+            dataU .*= transpose(sqrt.(sum(abs2, trj,dims=1)))
+    elseif density_compensation == :none
+        # do nothing here
+    elseif isa(density_compensation, Symbol)
+        error("`density_compensation` can only be `:radial_3D`, `:radial_2D`, `:none`, or of type  `AbstractVector{<:AbstractVector{T}}`")
+    else
+        @simd for i ∈ CartesianIndices(dataU)
+            dataU .*= density_compensation
+        end
+    end
+end
+
+function calcFilteredBackProjection(data::AbstractArray{Complex{T},3}, trj::AbstractVector{<:AbstractMatrix{T}}, U::AbstractMatrix{Complex{T}}, img_shape::NTuple{N,Int}; density_compensation::Union{Symbol, <:AbstractVector{<:AbstractVector{T}}}=:radial_3D, verbose = false) where {N,T}
     p = plan_nfft(reduce(hcat,trj), img_shape; precompute=TENSOR, blocking = true, fftflags = FFTW.MEASURE)
+    Ncoils = size(data,3)
     xbp = Array{Complex{T}}(undef, img_shape..., Ncoils)
 
     dataU = similar(@view data[:,:,1]) # size = Ncycles*Nr x Nt
     img_idx = CartesianIndices(img_shape)
     t = @elapsed for icoil ∈ axes(data,3)
-        if density_compensation == :radial_3D
-            @simd for i ∈ CartesianIndices(dataU)
-                dataU[i] = data[i,icoil] * conj(U[i[2],1]) * sum(abs2, @view trj[i[2]][:,i[1]])
-            end
-        elseif density_compensation == :radial_2D
-            @simd for i ∈ CartesianIndices(dataU)
-                dataU[i] = data[i,icoil] * conj(U[i[2],1]) * sqrt(sum(abs2, @view trj[i[2]][:,i[1]]))
+        dataU = data[:,:,icoil] .* U[:,1]'
+        for it in axes(data,2)
+            if typeof(density_compensation) <:AbstractVector
+                @views applyDensityCompensation!(dataU[:,it],trj[it]; density_compensation = density_compensation[it])
+            else
+                @views applyDensityCompensation!(dataU[:,it],trj[it]; density_compensation)
             end
-        elseif density_compensation == :none
-            # no density compensation; premultiply data with inverse of sampling density before calling function
-            @simd for i ∈ CartesianIndices(dataU)
-                dataU[i] = data[i,icoil] * conj(U[i[2],1])
-            end
-        elseif isa(density_compensation, Symbol)
-            error("`density_compensation` can only be `:radial_3D`, `:radial_2D`, `:none`, or of type  `AbstractVector{<:AbstractVector{T}}`")
+        end
```

Comment:
Would it make sense to move this for-loop into the `applyDensityCompensation!` function?